src/libutil/c/nix_api_util.h

```diff
@@ -0,0 +1,290 @@
+ *  @{
+ */
+/**
+ * @brief Retrieves a setting from the nix global configuration.
+ *
+ * This function requires nix_libutil_init() to be called at least once prior to
+ * its use.
+ *
+ * @param[out] context optional, Stores error information
+ * @param[in] key The key of the setting to retrieve.
+ * @param[out] value A pointer to a buffer where the value of the setting will
+ * be stored.
+ * @param[in] n The size of the buffer pointed to by value.
+ * @return NIX_ERR_KEY if the setting is unknown, NIX_ERR_OVERFLOW if the
+ * provided buffer is too short, or NIX_OK if the setting was retrieved
+ * successfully.
+ */
+nix_err nix_setting_get(nix_c_context * context, const char * key, char * value, int n);
+
+/**
+ * @brief Sets a setting in the nix global configuration.
+ *
+ * Use "extra-<setting name>" to append to the setting's value.
+ *
+ * Settings only apply for new State%s. Call nix_plugins_init() when you are
+ * done with the settings to load any plugins.
+ *
+ * @param[out] context optional, Stores error information
+ * @param[in] key The key of the setting to set.
+ * @param[in] value The value to set for the setting.
+ * @return NIX_ERR_KEY if the setting is unknown, or NIX_OK if the setting was
+ * set successfully.
+ */
+nix_err nix_setting_set(nix_c_context * context, const char * key, const char * value);
+
+/**
+ *  @}
+ */
+// todo: nix_plugins_init()
+
+/**
+ * @brief Retrieves the nix library version.
+ *
+ * Does not fail.
+ * @return A static string representing the version of the nix library.
+ */
+const char * nix_version_get();
+
+/** @addtogroup errors
+ *  @{
+ */
+/**
+ * @brief Retrieves the most recent error message from a context.
+ *
+ * @pre This function should only be called after a previous nix function has
+ * returned an error.
+ *
+ * @param[out] context optional, the context to store errors in if this function
+ * fails
+ * @param[in] ctx the context to retrieve the error message from
+ * @param[out] n optional: a pointer to an unsigned int that is set to the
+ * length of the error.
+ * @return nullptr if no error message was ever set,
+ *         a borrowed pointer to the error message otherwise.
+ */
+const char * nix_err_msg(nix_c_context * context, const nix_c_context * ctx, unsigned int * n);
+
+/**
+ * @brief Retrieves the error message from errorInfo in a context.
+ *
+ * Used to inspect nix Error messages.
+ *
+ * @pre This function should only be called after a previous nix function has
+ * returned a NIX_ERR_NIX_ERROR
+ *
+ * @param[out] context optional, the context to store errors in if this function
+ * fails
+ * @param[in] read_context the context to retrieve the error message from
+ * @param[out] value The allocated area to write the error string to.
+ * @param[in] n Maximum size of the returned string.
+ * @return NIX_OK if there were no errors, an error code otherwise.
+ */
+nix_err nix_err_info_msg(nix_c_context * context, const nix_c_context * read_context, char * value, int n);
```

Comment:
I think the docs should be more explicit on how this differs from `nix_err_msg` in output. Besides, the API is somewhat tricky to use, consider this:

When this function is called initially, the caller has no idea how large of a buffer to allocate. If the function fails, it will have to reallocate a larger buffer, and try again. In at leat Go and Rust the returned string will then further have to be copied over into its final form (e.g. Rust enforces the UTF* correctness on that step, and Go casts the array to an actual String by copying).

Instead, it could be helpful if this API returned the value as a `const char*`. It won't overlive the context, and that's a fair assumption given most call sites will have to copy immediately after either way. Alternatively, consider a case where it takes `*int` instead and if the `value` is NULL, it will **write** the actual length out.

Reply:
So you're suggesting the library to allocate the string and pass ownership to the caller?

Reply:
I suggest the library to return `const char*` much like `nix_err_msg` does, with a note that the value is valid until the next library call within this thread (I believe it's effectively bound to the context, lifetime, is it not?)